### PR TITLE
Add OTEL_EXPORTER environment variable

### LINKED ISSUE #710: Add OTEL_EXPORTER environment variable
**What are you trying to achieve?**

To simplify the configuration of auto-instrumentations via environment variables, especially for first time users, add new variable `OTEL_EXPORTER` as a way to configure a list of exporters to use, e.g. `OTEL_EXPORTER=zipkin,prometheus,otlp_metrics`

**Additional context.**
This issue assumes that #666 is merged and we have environment variables recommendation in spec.

### PR DESCRIPTION
**Pull request by trask**
Fixes #710

## Changes

Adds `OTEL_EXPORTER` environment variable to select the exporter to be used.

Adds `OTEL_EXPORTER_OTLP_ENDPOINT` environment variable to configure the ingest endpoint for both OTLP spans and metrics.

## Notes

1. I added `OTEL_EXPORTER_OTLP_ENDPOINT` in this PR because I think it makes the default value `OTEL_EXPORTER=otlp` make more sense.

   (It also happens to be the env var that the Java SDK uses currently, and seems simpler for most OTLP users, who will be sending spans and metrics to the same endpoint)

   An alternative would be to specify the default value as `OTEL_EXPORTER=otlp_span,otlp_metric`.

2. I went with the singular `OTEL_EXPORTER` over the plural `OTEL_EXPORTERS`, since the common use case is to configure just one exporter and so this seems less confusing.

### CODE REVIEW

File: specification/sdk-environment-variables.md [hunk 1]
```diff
@@ -49,6 +55,12 @@ The goal of this specification is to unify the environment variable names betwee
 | ----------------------------- | -------------------------- | ------------------------------------------------------------------------------------------------------------ |
 | OTEL_EXPORTER_ZIPKIN_ENDPOINT | Endpoint for Zipkin traces | <!-- markdown-link-check-disable --> "http://localhost:9411/api/v2/spans"<!-- markdown-link-check-enable --> |
 
+## Exporter Selection
+
+| Name          | Description                                                                  | Default | Notes                                                              |
+| ------------- | ---------------------------------------------------------------------------- | ------- | ------------------------------------------------------------------ |
+| OTEL_EXPORTER | Exporter to be used, can be a comma-separated list to use multiple exporters | "otlp"  | Other known values: "jaeger", "zipkin", "otlp_span", "otlp_metric" |
```
Comment (anuraaga) [inline specification/sdk-environment-variables.md:62]: prometheus?

**Review by anuraaga — APPROVED** [review]

File: specification/sdk-environment-variables.md (hunk 1 above)
Comment (iNikem) [inline specification/sdk-environment-variables.md:62]: How can I specify different exporters for spans and metrics? E.g. OTLP for spans, but Prometheus for metrics?

File: specification/sdk-environment-variables.md [hunk 2]
```diff
@@ -19,19 +19,25 @@ The goal of this specification is to unify the environment variable names betwee
...
 ## OTLP Span Exporter
 
-| Name                             | Description                                | Default |
-| -------------------------------- | ------------------------------------------ | ------- |
-| OTEL_EXPORTER_OTLP_SPAN_TIMEOUT  | Max waiting time to export each span batch | -       |
-| OTEL_EXPORTER_OTLP_SPAN_ENDPOINT | Ingest endpoint for OTLP spans             | -       |
+| Name                             | Description                                | Default | Notes                                              |
+| -------------------------------- | ------------------------------------------ | ------- | -------------------------------------------------- |
+| OTEL_EXPORTER_OTLP_SPAN_TIMEOUT  | Max waiting time to export each span batch | -       |                                                    |
+| OTEL_EXPORTER_OTLP_SPAN_ENDPOINT | Ingest endpoint for OTLP spans             | -       | Overrides OTEL_EXPORTER_OTLP_ENDPOINT if non-empty |
```
Comment (iNikem) [inline specification/sdk-environment-variables.md:33]: Should default value be localhost and default OTLP port in the collector?

File: specification/sdk-environment-variables.md (hunk 1 above)
Comment (carlosalberto) [reply specification/sdk-environment-variables.md:62]: Probably move the list of known values outside the table? Could be handy in case we add more ;)

Comment (codeboten) [reply specification/sdk-environment-variables.md:62]: Would it make sense to support `OTEL_EXPORTER_SPAN` `OTEL_EXPORTER_METRIC` for differentiating between the two here with the default for both to be `otlp`?

File: specification/sdk-environment-variables.md [hunk 3]
```diff
@@ -19,19 +19,25 @@ The goal of this specification is to unify the environment variable names betwee
 | OTEL_BSP_MAX_QUEUE_SIZE        | Maximum queue size                             | 2048    |                                                       |
 | OTEL_BSP_MAX_EXPORT_BATCH_SIZE | Maximum batch size                             | 512     | Must be less than or equal to OTEL_BSP_MAX_QUEUE_SIZE |
 
+## OTLP Exporters
+
+| Name                        | Description                                     | Default |
+| --------------------------- | ----------------------------------------------- | ------- |
+| OTEL_EXPORTER_OTLP_ENDPOINT | Ingest endpoint for both OTLP spans and metrics | -       |
```
Comment (codeboten) [inline specification/sdk-environment-variables.md:26]: It's worth noting that there are additional configuration options defined here: https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/protocol/exporter.md

**Review by codeboten — COMMENTED** [review]
Is there a mention in the spec as to what the known values that should be used to identify exporters are and how exporters should register these values? That would be useful for anyone implementing additional exporters.

**Comment by SergeyKanzhelev** [comment]
> Adds `OTEL_EXPORTER` environment variable to select the exporter to be used.

would it make more sense for every exporter to keep `_ENABLED` variable (like `OTEL_EXPORTER_OTLP_SPAN_ENABELD` and `OTEL_EXPORTER_JAEGER_ENABLED`)? This way we can enable any number of exporters and no need to keep the centralized list. Also it is more extensible as the list of exporters can be extended in very logical way.

Having `OTEL_EXPORTER` is more aligned with the existence of a single batch processor setting. But I'd argue it's more logical to have enable flag configured with the exporter.

**Comment by trask** [comment]
There are two alternatives suggested above. I like both of them, except around handling of the default value.

1. Suggestion: split into two env vars, one for span exporter (`OTEL_EXPORTER_SPAN`) and one for metric exporter (`OTEL_EXPORTER_METRIC`)

   The default in this case would be `OTEL_EXPORTER_SPAN=otlp` and `OTEL_EXPORTER_METRIC=otlp`.

   If someone wants to use jaeger, would they need to set both `OTEL_EXPORTER_SPAN=jaeger` and `OTEL_EXPORTER_METRIC=none`?

2. Suggestion: split out into a boolean flag for each exporter (`OTEL_EXPORTER_*_ENABLED`)

   The default in this case would be `OTEL_EXPORTER_OTLP_ENABLED=true` and all others `OTEL_EXPORTER_*_ENABLED=false`.

   If someone wants to use jaeger, would they need to set both `OTEL_EXPORTER_JAEGER_ENABLED=true` and `OTEL_EXPORTER_OTLP_ENABLED=false`.

Wondering if anyone has suggestions how to handle default value better in either of these cases? Or should we stick with the single environment variable `OTEL_EXPORTER` with default value `otlp`?

File: specification/sdk-environment-variables.md (hunk 3 above)
Comment (trask) [reply specification/sdk-environment-variables.md]: thx, I hadn't seen that!

File: specification/sdk-environment-variables.md (hunk 2 above)
Comment (trask) [reply specification/sdk-environment-variables.md]: now pointing this section to https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/protocol/exporter.md which has these default values 👍

File: specification/sdk-environment-variables.md (hunk 1 above)
Comment (trask) [reply specification/sdk-environment-variables.md]: done

Comment (trask) [reply specification/sdk-environment-variables.md]: @iNikem `OTEL_EXPORTER=otlp_span,prometheus`

@codeboten see comment below: https://github.com/open-telemetry/opentelemetry-specification/pull/943#issuecomment-692449625

File: specification/sdk-environment-variables.md [hunk 4]
```diff
@@ -49,6 +39,14 @@ The goal of this specification is to unify the environment variable names betwee
 | ----------------------------- | -------------------------- | ------------------------------------------------------------------------------------------------------------ |
 | OTEL_EXPORTER_ZIPKIN_ENDPOINT | Endpoint for Zipkin traces | <!-- markdown-link-check-disable --> "http://localhost:9411/api/v2/spans"<!-- markdown-link-check-enable --> |
 
+## Exporter Selection
+
+| Name          | Description                                                                  | Default |
+| ------------- | ---------------------------------------------------------------------------- | ------- |
+| OTEL_EXPORTER | Exporter to be used, can be a comma-separated list to use multiple exporters | "otlp"  |
+
+Known values for OTEL_EXPORTER are: "otlp", "jaeger", "zipkin", "prometheus", "otlp_span", "otlp_metric".
```
Comment (iNikem) [inline specification/sdk-environment-variables.md:48]: Probably should explicitly say that `otlp` means `otlp_span+otlp_metric`?

**Review by iNikem — APPROVED** [review]

**Comment by iNikem** [comment]
I prefer the third alternative, which is currently written in PR :)

File: specification/sdk-environment-variables.md (hunk 4 above)
Comment (trask) [reply specification/sdk-environment-variables.md:48]: done

**Comment by trask** [comment]
In case we go with `OTEL_EXPORTER`, and because I ❤️ naming, does anyone care between:
* `OTEL_EXPORTER=otlp_span`
* `OTEL_EXPORTER=otlp-span`
* `OTEL_EXPORTER=otlpspan`

Exporter names seem sort of like propagator names, not sure if there's any convention we can follow from those?

**Comment by iNikem** [comment]
For some reason `OTEL_EXPORTER=otlp_span` seems the most natural to me

**Comment by SergeyKanzhelev** [comment]
@trask you got me on default values. Single variable is better from this perspective.

As for naming, I again tend to lean to separate variables for metrics and spans. It feels like two separate pieces of configuration semantically.

Since I also ❤️  naming, should we get PM to this PR to help nail the desired semantic of this config settings ;-)?

**Review by carlosalberto — APPROVED** [review]

**Review by codeboten — APPROVED** [review]
Thanks for the updates, this LGTM

**Comment by trask** [comment]
> (edit: I assumed "because I ❤️ naming" was a sarcasm. If not, than mine is =))

Mine was intentionally ambiguous 😂. I have a complicated relationship with naming.

**Review by SergeyKanzhelev — APPROVED** [review]

**Review by tigrannajaryan — APPROVED** [review]
LGTM

**Comment by SergeyKanzhelev** [comment]
Last edit was a day ago, so plan to merge tomorrow morning unless somebody will comment with blocking concerns

File: specification/sdk-environment-variables.md (hunk 4 above)
Comment (owais) [reply specification/sdk-environment-variables.md:48]: What about 3rd party or contrib exporters? This is language specific but spec should probably mention if we want to support that. 

For example, Python could support something like `OTEL_EXPORTER=otlp_span,my_exporter.MetricExporter'. Any values present in the list other than the well-known ones mentioned in the spec will have to be fully-qualified import paths to exporter implementations. Another issue this brings up is how do we tell if a custom exporter exports metrics, spans or logs. 

1. Each implementation could use some sort of type-checking to verify what kind of exporter something implements. If something implements multiple exporters, do we instantiate a single instance for all supported types or multiple instances per type?

2. Go with the other suggestion to have one env var per type i.e, `OTEL_SPAN_EXPORTER`, `OTEL_METRIC_EXPORTER` etc. 

Java today supports using an external exporter  by pointing to a .jar file `-Dotel.exporter.jar=path/to/external-exporter.jar`.  https://github.com/open-telemetry/opentelemetry-java-instrumentation#getting-started

I also have an open PR for Python that supports specifying import paths to load custom exporters.

**Comment by owais** [comment]
Other than OTLP, most other exporters expect an explicit value for "service name". OTLP expects a `service_name` resource attribute. The spec already specifies a way to configure resource attributes globally for a service using environment variables (`OTEL_RESOURCE_ATTRIBUTES`) so OTLP is covered but this leaves other exporters without a way to set the service name. We can either add a common `OTEL_SERVICE_NAME` env var or one env var per exporter like `OTEL_EXPORTER_JAEGER_SERVICE_NAME`.

File: specification/sdk-environment-variables.md [hunk 5]
```diff
@@ -49,6 +39,16 @@ The goal of this specification is to unify the environment variable names betwee
...
+
```
Comment (owais) [inline specification/sdk-environment-variables.md:49]: What about 3rd party or contrib exporters? This is language specific but spec should probably mention if we want to support that.

For example, Python could support something like `OTEL_EXPORTER=otlp_span,my_exporter.MetricExporter'. Any values present in the list other than the well-known ones mentioned in the spec will have to be fully-qualified import paths to exporter implementations. Another issue this brings up is how do we tell if a custom exporter exports metrics, spans or logs.

    Each implementation could use some sort of type-checking to verify what kind of exporter something implements. If something implements multiple exporters, do we instantiate a single instance for all supported types or multiple instances per type?

    Go with the other suggestion to have one env var per type i.e, OTEL_SPAN_EXPORTER, OTEL_METRIC_EXPORTER etc.

Java today supports using an external exporter by pointing to a .jar file -Dotel.exporter.jar=path/to/external-exporter.jar. https://github.com/open-telemetry/opentelemetry-java-instrumentation#getting-started

I also have an open PR for Python that supports specifying import paths to load custom exporters.

**Comment by trask** [comment]
> We can either add a common `OTEL_SERVICE_NAME` env var or one env var per exporter like `OTEL_EXPORTER_JAEGER_SERVICE_NAME`.

Hey @owais, check out (and vote for!) #709

File: specification/sdk-environment-variables.md (hunk 5 above)
Comment (trask) [reply specification/sdk-environment-variables.md:49]: I'd like to suggest that custom exporters are out of scope for this PR, can you file a separate issue/PR for this?

Comment (owais) [reply specification/sdk-environment-variables.md:49]: What this PR suggest does not close the door for supporting custom exporters later so dealing with that in a separate PR sounds good to me.

**Review by cijothomas — APPROVED** [review]

**Pull request merged**